### ISSUE TITLE
non_camel_case_types does not trigger for screaming single-word

### ISSUE THREAD

**Issue by jonhoo** [rust-lang/rust#116336]
I tried this code:

```rust
#![deny(non_camel_case_types)]

pub struct FOO;

pub enum BAR {
    BAZ,
}

fn main() {}
```

I expected to see this happen: `FOO`, `BAR`, and `BAZ` should all trigger the `non_camel_case_types` lint (at least assuming it follows [RFC430](https://rust-lang.github.io/rfcs/0430-finalizing-naming-conventions.html), and so the code should not compile.

Instead, this happened: The code compiled with no warnings. All three cases start triggering the lint if something like `_X` is added to them.

### Meta

`rustc --version --verbose`:
```
rustc 1.73.0-beta.4 (9f37cd4f7 2023-09-01)
binary: rustc
commit-hash: 9f37cd4f79d289492ccaa8f28733b765498c95e5
commit-date: 2023-09-01
host: aarch64-apple-darwin
release: 1.73.0-beta.4
LLVM version: 17.0.0
```

**Comment by kadiwa4** [comment]
Related: #60570, which describes drawbacks of always disallowing uppercase type names

**Comment by joshtriplett** [comment]
https://github.com/rust-lang/rust/issues/60570#issuecomment-1812917719
